### ISSUE TITLE
Unable to tune on GSM8K?

### ISSUE THREAD

**Issue by PanDuTR** [gepa-ai/gepa#170]
I'm currently attempting to fine-tune a prompt for the **GSM8K** dataset using a local instance of **Qwen3-4B-Instruct** as both the *task language model (task_lm)* and the *reflective language model (reflective_lm)*.  

Despite using large training and validation sets (in the thousands of samples), I consistently observe the following messages during training:  
> - `"All subsample scores perfect. Skipping."`  
> - `"Reflective mutation did not propose a new candidate."`  

These indicate that the reflective reasoning process isn’t learning effectively.

Given that Qwen3-4B-Instruct already achieves ~0.85 accuracy on GSM8K without any prompt tuning, I suspect the current setup may not be conducive to effective learning via **GEPA** (Generalized Prompt Engineering via Reflection and Adaptation).  

With that, I have a few questions:  
1. Does the reflective_lm need to be stronger (in capability or performance) than the task_lm for GEPA to work effectively? Or  in other words - self-reflection would not work for GEPA?
2. If both the task_lm and the seed prompt are already strong (e.g., achieving 0.85 accuracy on GSM8K without tuning), would GEPA still yield meaningful improvements? In particular, would using the *same model* for both task_lm and reflective_lm result in any meaningful reflective behavior — or would it simply degenerate into trivial or unproductive iterations?  
3. How sensitive is the tuning process to the seed prompts?
4. Should I instead investigate potential issues in my code, hyperparameters, or configuration?

Any insights or recommendations would be greatly appreciated — especially 
regarding model selection, seed prompt design, or training setup.

**Comment by LakshyAAAgrawal** [comment]
1. All the experiments in the GEPA paper (https://arxiv.org/abs/2507.19457) are based on self-reflection, so it does work very well. Having said that, we do recommend users to provide the strongest available / frontier models for reflection, as that enables much better performance. However, in this case the issue does not seem to issue from reflection, since most of the reflection steps are being skipped due to the signal being very sporadic (almost all successes, very few failures).
2. Yes, I believe so, but we need a mechanism for GEPA to be able to select a mix of successful and failing examples. In every iteration, GEPA stochastically samples a minibatch of examples from the trainset, evaluates them, and if at least one of them is failing, then tries to improve on this minibatch. What is happening right now is, in almost all iterations, it is just getting succesful cases, and skipping the iteration. One way to address this would be to stratify sample your trainset, include all failing examples from the trainset, with 50:50 ratio of successful examples. This will avoid the issue. You can also look at the BatchSampler argument in GEPA ([class BatchSampler(Protocol[DataId, DataInst]):](https://github.com/gepa-ai/gepa/blob/main/src/gepa/strategies/batch_sampler.py#L13) and https://github.com/gepa-ai/gepa/blob/main/src/gepa/api.py#L50), that guides GEPA in selecting the next round's minibatch.
3. The only expectation from the seed prompt is for it to specify the task in sufficient detail, and act as a reasonable starting point. However, since the evolution tries to rewrite the seed prompt, it often is sensitive to the seed.
4. I would advise looking at sampling from the trainset and looking into the batch sampler argument.

**Comment by LakshyAAAgrawal** [comment]
I am closing the issue for now, but please feel free to reopen it for any further discussion and I would be glad to know about how your experiments go.

**Comment by LakshyAAAgrawal** [comment]
You can also consider increasing the reflection minibatch size. Given that the success rate is high, having a large stochastically selected minibatch will have a better chance of selecting at least one failing example.

**Comment by PanDuTR** [comment]
Thank you for the prompt response, the suggestions all make sense, will let you know how it goes!
